lib/pbench/cli/agent/commands/tools/list.py

```diff
@@ -36,35 +45,45 @@ def execute(self):
             host_tools = {}
             for group in groups:
                 host_tools[group] = {}
-                for path in self.gen_tools_group_dir(group).glob("*/**"):
-                    if self.context.with_option:
-                        host_tools[group][path.name] = [
-                            p.read_text() for p in self.tools(path)
-                        ]
-                    else:
-                        host_tools[group][path.name] = [p for p in self.tools(path)]
+                try:
+                    for path in self.gen_tools_group_dir(group).glob("*/**"):
+                        if self.context.with_option:
+                            host_tools[group][path.name] = [
+                                p.read_text() for p in self.tools(path)
+                            ]
+                        else:
+                            host_tools[group][path.name] = [p.name for p in self.tools(path)]
+                except BadToolGroup:
+                    self.logger.error("Bad tool group name: %s", group)
+                    return 1
             if host_tools:
                 for k, v in host_tools.items():
                     for h, t in v.items():
-                        print("%s: %s %s" % (k, h, t))
+                        print(f"{k}: {h} {t}")
         else:
             # List the groups which include this tool
             group_list = []
             for group in groups:
-                tg_dir = self.gen_tools_group_dir(group)
-                if not tg_dir.exists():
-                    self.logger.error("bad or missing tool group %s", group)
+                try:
+                    tg_dir = self.gen_tools_group_dir(group)
+                    if not tg_dir.exists():
+                        self.logger.error("bad or missing tool group %s", group)
+                        status = 1
+                        continue
+                except BadToolGroup:
+                    self.logger.error("Bad Tool Group: %s", group)
```

Comment:
It seems like we'd never hit line 70, but this exception.  So perhaps we should move that message here, and report missing tool group there with a status code of `2`.

Reply:
So the  exit codes should be: bad tool group = 1, missing tool group = 2? What about a bad tool name? Or a bad setup (the no pbench_run directory found case)?

Reply:
I would vote against having separate codes for each sort of error, unless we expect the invoker to take explicit actions based on the return code.

I think the typical convention is to return `2` to indicate problems invoking the command (like bad or conflicting options) and to return `1` to indicate that the intended effect of the command failed for reasons other than its invocation.